tests/robustness/multiclient_test/framework/harness.go

```diff
@@ -0,0 +1,241 @@
+	"github.com/kopia/kopia/tests/tools/kopiarunner"
+)
+
+const (
+	dataSubPath     = "robustness-data"
+	metadataSubPath = "robustness-metadata"
+)
+
+var repoPathPrefix = flag.String("repo-path-prefix", "", "Point the robustness tests at this path prefix")
+
+// NewHarness returns a test harness.
+func NewHarness(ctx context.Context) *TestHarness {
+	dataRepoPath := path.Join(*repoPathPrefix, dataSubPath)
+	metadataRepoPath := path.Join(*repoPathPrefix, metadataSubPath)
+
+	th := &TestHarness{}
+	th.init(ctx, dataRepoPath, metadataRepoPath)
+
+	return th
+}
+
+// TestHarness provides a Kopia robustness.Engine.
+type TestHarness struct {
+	dataRepoPath string
+	metaRepoPath string
+
+	baseDirPath string
+	fileWriter  *MultiClientFileWriter
+	snapshotter *MultiClientSnapshotter
+	persister   *snapmeta.KopiaPersister
+	engine      *engine.Engine
+
+	skipTest bool
+}
+
+func (th *TestHarness) init(ctx context.Context, dataRepoPath, metaRepoPath string) {
+	th.dataRepoPath = dataRepoPath
+	th.metaRepoPath = metaRepoPath
+
+	// Override ENGINE_MODE env variable. Multiclient tests can only run in SERVER mode.
+	log.Printf("Setting %s to %s\n", snapmeta.EngineModeEnvKey, snapmeta.EngineModeServer)
+
+	err := os.Setenv(snapmeta.EngineModeEnvKey, snapmeta.EngineModeServer)
+	if err != nil {
+		log.Printf("Error setting ENGINE_MODE to server: %s", err.Error())
+		os.Exit(1)
+	}
+
+	// the initialization state machine is linear and bails out on first failure
+	if th.makeBaseDir() && th.getFileWriter() && th.getSnapshotter() &&
+		th.getPersister() && th.getEngine(ctx) {
+		return // success!
+	}
+
+	err = th.Cleanup(ctx)
+	if err != nil {
+		log.Printf("Error cleaning up the engine: %s\n", err.Error())
+		os.Exit(2)
+	}
+
+	if th.skipTest {
+		os.Exit(0)
+	}
+
+	os.Exit(1)
+}
+
+func (th *TestHarness) makeBaseDir() bool {
+	baseDir, err := ioutil.TempDir("", "engine-data-")
+	if err != nil {
+		log.Println("Error creating temp dir:", err)
+		return false
+	}
+
+	th.baseDirPath = baseDir
+
+	return true
+}
+
+func (th *TestHarness) getFileWriter() bool {
+	fw, err := NewMultiClientFileWriter(
+		func() (FileWriter, error) { return fiofilewriter.New() },
+	)
+	if err != nil {
+		if errors.Is(err, fio.ErrEnvNotSet) {
+			log.Println("Skipping robustness tests because FIO environment is not set")
+
+			th.skipTest = true
+		} else {
+			log.Println("Error creating fio FileWriter:", err)
+		}
+
+		return false
+	}
+
+	th.fileWriter = fw
+
+	return true
+}
+
+func (th *TestHarness) getSnapshotter() bool {
+	server, err := snapmeta.NewSnapshotter(th.baseDirPath)
+	if err != nil {
+		if errors.Is(err, kopiarunner.ErrExeVariableNotSet) {
+			log.Println("Skipping robustness tests because KOPIA_EXE is not set")
+
+			th.skipTest = true
+		} else {
+			log.Println("Error creating kopia Snapshotter:", err)
+		}
+
+		return false
+	}
+
+	newClientFn := func(baseDirPath string) (ClientSnapshotter, error) {
+		return snapmeta.NewSnapshotter(th.baseDirPath)
+	}
```

Comment:
Q: It appears from this that all clients make snapshots of the same base directory. Am I correct in my understanding?

Reply:
This `baseDirPath` gets passed all the way down to [kopiarun.go:37](https://github.com/kastenhq/kopia/blob/c6dc82b7af02a20919b8fbbdf7f962cce9c8cb32/tests/tools/kopiarunner/kopiarun.go#L37) where it is used to create a unique config directory for each runner. It does not seem to be used anywhere else along the way.

Snapshots use the data directory that is defined in the `FileWriter`, which is unique for each client. See [action.go:160](https://github.com/kastenhq/kopia/blob/c6dc82b7af02a20919b8fbbdf7f962cce9c8cb32/tests/robustness/engine/action.go#L160). Restores either use that same data directory or use a unique temporary directory for each restore (depending on which engine action is specified).